GH-1912: GraphMem2

arne-bdt opened the pull request:
New in-memory, general-purpose, non-transactional graphs as successors of GraphMem: All variants strictly use term-equality and do not support Iterator#remove. (GraphMem uses value-equality for object nodes)

GraphMem2Legacy:
- Purpose: Use this graph implementation if you want to maintain the 'old' behavior of GraphMem or if your memory constraints prevent you from utilizing more memory-intensive solutions.
- Slightly improved performance compared to GraphMem
- Simplified implementation, primarily due to lack of support for Iterator#remove
- The heritage of GraphMem:
  - Same basic structure
  - Same memory consumption
- Also based on HashCommon

GraphMem2Fast:
- Purpose: GraphMem2Fast is a strong candidate for becoming the new default in-memory graph in the upcoming Jena 5, thanks to its improved performance and relatively minor increase in memory usage.
- Faster than GraphMem2Legacy (specially Graph#add, Graph#find and Graph#stream)
- Memory consumption is about 6-35% higher than GraphMem2Legacy
- Maps and sets are not based on HashCommon, but use a faster custom alternative (only #remove is a bit slower)
- Benefits from multiple small optimizations
- The heritage of GraphMem:
  - Also uses 3 hash-maps indexed by subjects, predicates, and objects
  - Values of the maps also switch from arrays to hash sets for the triples

GraphMem2Roaring
- Purpose: GraphMem2Roaring is ideal for handling extremely large graphs. If you frequently work with such massive data structures, this implementation could be your top choice.
- Graph#contains is faster than GraphMem2Fast
- Better performance than GraphMem2Fast for operations with triple matches for the pattern S_O, SP_, and _PO on large graphs, due to bit-operations to find intersecting triples
- Memory consumption is about 7-99% higher than GraphMem2Legacy
- Suitable for really large graphs like bsbm-5m.nt.gz, bsbm-25m.nt.gz, and possibly even larger
- Simple and straightforward implementation
- No heritage of GraphMem
- Internal structure:
  - One indexed hash set (same as GraphMem2Fast uses) that holds all triples
  - Three hash maps indexed by subjects, predicates, and objects with RoaringBitmaps as values
  - The bitmaps contain the indices of the triples in the central hash set

Other Changes:
- org.apache.jena.graph.test.TestGraph
  - added GraphMem2Fast, GraphMem2Legacy and GraphMem2Roaring to the suite
- GraphMem:
  - moved property "TripleStore store" from GraphMemBase to GraphMem --> needed this to make a clean GraphMem2, which also extends GraphMem but the TripleStore interface is slightly different.
- pom.xml:
  - added dependency roaringbitmap 0.9.44
- jena-benchmarks-jmh
  - added the three new graph implementations to the benchmarks
  - randomized the order of test data in some benchmarks to prevent them from showing order dependent behaviour
  - added benchmarks for sets and maps comparing
    - HashCommonSet vs. FastHashSet vs. Java HashSet
    - HashCommonMap vs. FastHashMap vs. Java HashMap

GitHub issue resolved #1912

----

 - [X] Tests are included.
 - [X] Documentation change and updates are provided for the [Apache Jena website](https://github.com/apache/jena-site/)
 - [X] Commits have been squashed to remove intermediate development commit messages.
 - [X] Key commit messages start with the issue number (GH-xxxx, or if in JIRA, JENA-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

afs commented:
PR builds.
New dependency (roaringbitmaps) is Apache 2.0 licensed.

Looking good!

afs commented:
I tried 50M and 100m BSBM using the async parser to keep the parser work on another thread.

The data files wre on an NVMe SSD.
Heap size: -Xmx40G

<details>
<summary>Results for BSBM 50m</summary>

```
**** GraphMem2Roaring ****
**** Data: bsbm-50m.nt.gz ****
Load time  = 116.362 s
Load rate  = 429,742 TPS
Load space = 15,970M 

**** GraphMem2Fast ****
**** Data: bsbm-50m.nt.gz ****
Load time  = 117.762 s
Load rate  = 424,633 TPS
Load space = 16,347M 

**** GraphMem2Legacy ****
**** Data: bsbm-50m.nt.gz ****
Load time  = 120.876 s
Load rate  = 413,694 TPS
Load space = 18,169M 

**** GraphMem ****
**** Data: bsbm-50m.nt.gz ****
Load time  = 121.770 s
Load rate  = 410,656 TPS
Load space = 20,081M 
```

</details>

and

<details><summary>Results for BSBM 100m</summary>

```
**** GraphMem2Roaring ****
**** Data: bsbm-100m.nt.gz ****
Load time  = 235.356 s
Load rate  = 424,891 TPS
Load space = 31,733M 

**** GraphMem2Fast ****
**** Data: bsbm-100m.nt.gz ****
Load time  = 238.313 s
Load rate  = 419,619 TPS
Load space = 32,143M 

**** GraphMem2Legacy ****
**** Data: bsbm-100m.nt.gz ****
Load time  = 249.471 s
Load rate  = 400,851 TPS
Load space = 35,811M 

**** GraphMem ****
**** Data: bsbm-100m.nt.gz ****
Load time  = 259.397 s
Load rate  = 385,512 TPS
Load space = 39,582M 
```

</details>

arne-bdt commented:
To exclude the parsing part from the benchmarks, I usually parse the triples directly into an ArrayList first:
```
public List<Triple> readTriples(String graphUri) {
        var list = new ArrayList<Triple>();
        var g1 = new GraphMem() {
            @Override
            public void add(Triple t) {
                list.add(t);
            }
        };
        RDFDataMgr.read(g1, graphUri);
        return list;
    }
```

afs commented:
I ran the Jena core test suite just "because".

For the GraphMem2, graphs there are 16 errors and 13 failures. Most errors look to all be lack of iterator.remove and the failures are mostly value/term related.  No surprises.

There some test that explicitly test for GraphMemBase - SimpleGraphMaker has some reference counting of (create,use)/close pairing. GraphMaker should probably go or at least migrate to the ontology code where and not be used elsewhere.

GraphMem2Roaring needed one tweek. In RoaringBitmapTripleIterator,ext there is a CME test. With that, there were 40 errors. When I removed it, 16 errors like the others.

I also tried a graph wrappers to simulate term semantics on GraphMem and also one that simulated value semantics on GraphMem2Fast. Similar sort of errors and failures plus a few where test code was testing for GraphMem.

All these are FYI/later. Looks pretty good.

afs commented:
> To exclude the parsing part from the benchmarks, I usually parse the triples directly into an ArrayList first:

I wanted to test the load-to-graph use pattern.  I gave the load-to-list a go, and the speed is well in excess of the parser which is as it should be (GraphMem2Faster 2.9 M triples/s vs GraphMem 1M triples/s). The heap size for 100m BSBM falls in the deadzone of 32-40G where the runtime switches from compressed pointers to 64 bit pointers. BSBM has significant numbers of long literals.

afs commented on jena-core/src/main/java/org/apache/jena/mem2/store/fast/FastArrayBunch.java:
```diff
@@ -0,0 +1,190 @@
+
+    public abstract boolean areEqual(final Triple a, final Triple b);
+
+    public boolean containsKey(Triple t) {
```
```suggestion
    @Override
    public boolean containsKey(Triple t) {
```

arne-bdt replied:
Fixed this and found other places, where I forgot @override.

afs commented on jena-core/src/main/java/org/apache/jena/mem2/collection/FastHashSet.java:
```diff
@@ -0,0 +1,111 @@
+        return tryAdd(key, key.hashCode());
+    }
+
+    public boolean tryAdd(K value, int hashCode) {
```
```suggestion
    @Override
    public boolean tryAdd(K value, int hashCode) {
```

arne-bdt replied:
Fixed this and found other places, where I forgot @override.

afs commented:
The way forward:

I'll keep looking at the code as and when time permits - it looks good already.

There is some code tidying up : the idea is to make the code free of warnings for some 

When checking the code for releases, the no warnings/some warnings boundary is easy to see whereas a change in the number of warnings isn't so easy to notice. We don't have proscribed set of settings - I think I have close to the default in Eclipse. It does mean it isn't always the set of warnings that you use.

* Rewriting to make `@SuppressWarnings("squid:S1121")` and `"squid:2184"` unnecessary. 
* A few missing `@Override`
* Javadoc e.g. `@return` with no text.
* Use of generics in Javadoc..
* `Assert.assertThat` is deprecated. The one line fix for this gets rid of most of the warnings!

I've put in a code review - it doesn't matter if they are not all caught. They are be addressed later.

Then merge. 🥳

Then #1920 (GraphMemFactory)

Then put in factory calls into GraphMemFactory.

To me, that's enough to be ready 4.9.0. More access and further integration is a bonus but it's dependent on time available. I'd like to get 4.9.0 done and not wait unnecessarily. The northern vacation season approaches and it's push 5.x later.

At this point, exploring the consequences of term-semantics is much easier with GraphMem2* in the codebase. That'll include discovering long forgotten code like GraphMaker and its tight coupling to GraphMem1 - knocking off issues one by one.

arne-bdt commented:
Thanks.
I did my best to get rid of all the finds through IntelliJ code inspection and the Sonarqube plugin. All in default mode.
It seems that there are a few differences with Eclipse.
But I will check all the findings in the next two days. Today I don't have a computer within reach.

afs commented:
No rush - I can fix the warnings if needbe. 

I don't have Sonarqube on Jena, mainly because most contributors don't and I try to have the same experience as them.
(There again, maybe Jena5 is a time to change this and put IDE setup into the repo.)

arne-bdt commented on jena-core/src/main/java/org/apache/jena/mem2/iterator/NestedIterator.java:
```diff
@@ -0,0 +1,89 @@
+        while (this.parentIterator.hasNext()) {
+            this.currentIterator = this.mapper.apply(this.parentIterator.next());
+            if (this.currentIterator.hasNext()) {
+                return hasNext = true;
```
should be resolved now.

arne-bdt commented on jena-core/src/main/java/org/apache/jena/mem2/iterator/SparseArrayIterator.java:
```diff
@@ -0,0 +1,96 @@
+    public boolean hasNext() {
+        while (-1 < pos) {
+            if (null != entries[pos]) {
+                return hasNext = true;
```
should be resolved now.

arne-bdt commented on jena-core/src/main/java/org/apache/jena/mem2/spliterator/ArraySpliterator.java:
```diff
@@ -0,0 +1,106 @@
+    @Override
+    @SuppressWarnings("squid:S2184")
+    public long estimateSize() {
+        return pos - fromIndex;
```
I just removed the SuppressWarnings. Either way sonar lint does not like it.
With the (long) cast it says:
- Remove this unnecessary cast to "long".
- Casting '(pos - fromIndex)' to 'long' is redundant

arne-bdt commented on jena-core/src/main/java/org/apache/jena/mem2/iterator/NestedIterator.java:
```diff
@@ -0,0 +1,89 @@
+    @SuppressWarnings("squid:S1121")
+    public boolean hasNext() {
+        if (this.currentIterator.hasNext()) {
+            return hasNext = true;
```
I adopted your suggestion (not here but in my IDE)

arne-bdt commented on jena-core/src/main/java/org/apache/jena/mem2/spliterator/SparseArraySpliterator.java:
```diff
@@ -0,0 +1,114 @@
+    @Override
+    @SuppressWarnings("squid:S2184")
+    public long estimateSize() {
+        return pos - fromIndex;
```
I just removed the SuppressWarnings. Either way sonar lint does not like it.
With the cast it says:
- Remove this unnecessary cast to "long".
- Casting '(pos - fromIndex)' to 'long' is redundant

arne-bdt commented on jena-core/src/test/java/org/apache/jena/mem2/collection/AbstractJenaMapNodeTest.java:
```diff
@@ -0,0 +1,647 @@
+import java.util.stream.StreamSupport;
+
+import static org.apache.jena.testing_framework.GraphHelper.node;
+import static org.junit.Assert.*;
```
Should be fixed now.

arne-bdt commented on jena-core/src/main/java/org/apache/jena/mem2/iterator/NestedIterator.java:
```diff
@@ -0,0 +1,89 @@
+                return hasNext = true;
+            }
+        }
+        return hasNext = false;
```
I adopted your suggestion (not here but in my IDE)

arne-bdt commented on jena-core/src/main/java/org/apache/jena/mem2/store/legacy/LegacyTripleStore.java:
```diff
@@ -0,0 +1,152 @@
+ * Inner structure:
+ * - three {@link NodeToTriplesMapMem} instances for each of the three triple fields (subject, predicate, object)
+ * - each of these maps is a {@link HashCommonMap} with {@link Node} keys and {@link JenaSet<Triple>} values.
+ * - for up to 9 triples with the same subject, predicate or object, the {@link JenaSet<Triple>} is
```
I adopted your suggestion (not here but in my IDE)

arne-bdt commented:
I have tried to remedy all your findings.
Additionally:
- Added some missing Javadoc descriptions for parameters and return values.
- Tweaked the PatternClassifier a bit by testing `Triple#isConcrete` first.
- I found out that the RoaringBitmap BatchIterator is not that fast in the use cases covered by the benchmarks. So I replaced the implementation in RoaringBitmapTripleIterator.

I have no plans for further changes. I just ran the JMH benchmarks for more graphs and iterations (which often runs for many hours) and came across a few anomalies. That´s why I made changes to PatternClassifier  and RoaringBitmapTripleIterator.
I will still post the results here and highlight any remaining "anomalies". 

For example, org.apache.jena.mem.graph.TestGraphContainsTriple#graphContains almost always shows better performance of GraphMem2Fast than GraphMem2Legacy. With "bsbm-5m.nt.gz", however, GraphMem2Fast is clearly slower. I tried unsuccessfully to find the cause with the profiler. According to the profiler, GraphMem2Fast is faster.  I also tried to adjust the implementations:
- used the normal equals in containsKey in the FastArrayBunches
- set the maximum size of the FastArrayBunches to 8 elements (GraphMem2Legacy has 9)
- tried variants from HashCommonBase and the Java HashMap in FastHashBase for calculating the position from the hash code.
--> all without success. (so I reverted these experiments)

arne-bdt commented:
Found possible issue with RoaringBitmapIterator. Will look into it and submit fix soon.

afs commented:
I just going to merge this!

Merging this PR does not make the code unchangeable.

arne-bdt commented:
I think you are right.

It's just that I found that some iterators and streams were slower than my last PR for GraphMem iterators.
Sometimes I might be chasing phantoms. But there were other benchmarks with consistent behaviour.

So I decided to get to the bottom of it:
- In my first attempt, I only used the SparseArraySpliterator. Now it seems worth splitting it into SparseArraySpliterator and SparseArraySubSpliterator, as I did with GraphMem. The same goes for ArraySpliterator.
- the NestedIterator was a lame implementation. I replaced it with IteratorOfJenaSets.
- for RoaringBitmapIterator, using BufferedIterator was a good choice for #hasNext and #next, but not for #forEachRemaining. So I changed the implementation.

Some results in my benchmarks remain unclear to me:
- TestGraphContainsTriple for concrete triples on GraphMem2Fast shows faster performance than GraphMem for most graphs, but for some graphs (e.g. EQ and BSBM 5) the benchmarks consistently show lower performance.
- TestGraphFindAllWithForEachRemaining on GraphMem2Legacy proves faster than GraphMem for smaller graphs, but for larger graphs the benchmarks consistently show that GraphMem2Legacy is slightly slower. Due to their same inner structure, I would not expect hat. It might be an issue in the benchmarks but I am not sure. 

Apart from these points, I have not found anything else to worry about in the latest JMH benchmarks.

I have attached some of the latest JMH benchmarks for the various use cases. I only used 20 iterations to get the results faster, but they should show what you can expect from which implementation for different use cases.
[benchmark_results.zip](https://github.com/apache/jena/files/11878549/benchmark_results.zip)

arne-bdt commented:
The last commit had no changes in my code. (except for one missung license header)
The PR had conflicts with main, so I did a rebase and solved the conflicts.

arne-bdt commented:
Yes.
Though, I did not extend GraphFactoryMem to provide the new graphs.
Should I do that first?

afs commented:
Let's do the merge then make changes to GraphFactoryMem.

Long lived branches slowly become extra work to stay in-step, even small things like pom.xml.

Because Jena5 is an opportunity to make changes, I was hoping to improve/simplify/remove internals of jena-core as time permits. That doesn't directly affect GraphMem2* but it would be good to know nothing got touched by accident.

afs commented:
Merged!

I've set the GH and JIRA builds to run as preparation for 4.9.0.

I'll put in a PR for discussion for GraphMemFactory updates.
